### DISCUSSION TITLE
Type "str" cannot be assigned to type "str | None"

### DISCUSSION THREAD

**Discussion by emirkmo** [microsoft/pylance-release#3383] [Q&A] [answered]
Why can I not assign a dictionary of int: str to a variable that should be int: str | None.

I have a function that takes in this initial mapping which is int : str | None, and assigns a str for any None's. The initial dictionary is read from user input files so there may be None's. What am I missing?

Full error:
```Cannot assign member "markers" for type "Sites"
    "dict[int, str]" is incompatible with "dict[int, str | None]"
      TypeVar "_VT@dict" is invariant
        Type "str" cannot be assigned to type "str | None"
          Type cannot be assigned to type "None"```

**Comment by erictraut** [comment, marked as answer]
The type parameters for `dict` are invariant, so a `dict[int, str]` is not type compatible with a `dict[int, str | None]`. The type arguments must match exactly. This is necessary for type safety because `dict` is a mutable container. Consider the following.

```python
def func(a: dict[int, str]):
    b: dict[int, str | None] = a
    b['foo'] = None # Original dictionary `a` now contains a `None` value which violates its type!
```

For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).

**Comment by jackblk** [comment]
I have a similar issue and my solution is this:

```python
def func(a: dict[int, str] | dict[int, None]):
    # do your type check of the value here
```
